File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java
import java.util.*;

class Solution {
    public int[] solution(int rows, int columns, int[][] queries) {
        int[][] rc = new int[rows][columns];
        int[] answer = new int[queries.length];
        
        for (int j = 0; j < rows; j++) {
            for (int i = 0; i < columns; i++) {
                rc[j][i] = j * columns + i + 1;
            }
        }
        
        for (int k = 0; k < queries.length; k++) {
            int[] q = queries[k];
            int y1 = q[0] - 1;
            int x1 = q[1] - 1;
            int y2 = q[2] - 1;
            int x2 = q[3] - 1;
            
            List<Integer> change = new ArrayList<>();
            int temp = rc[y1][x1];
            change.add(temp);
            
            for (int x = x1; x < x2; x++) {
                int temp2 = temp;
                temp = rc[y1][x+1];
                rc[y1][x+1] = temp2;
                change.add(temp);
            }
            
            for (int y = y1; y < y2; y++) {
                int temp2 = temp;
                temp = rc[y+1][x2];
                rc[y+1][x2] = temp2;
                change.add(temp);
            }
            
            for (int x = x2; x > x1; x--) {
                int temp2 = temp;
                temp = rc[y2][x-1];
                rc[y2][x-1] = temp2;
                change.add(temp);
            }
            
            for (int y = y2; y > y1; y--) {
                int temp2 = temp;
                temp = rc[y-1][x1];
                rc[y-1][x1] = temp2;
                change.add(temp);
            }
            
            answer[k] = Collections.min(change);
            
        }
        
        return answer;
    }
}

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    rc = [[row * columns + column + 1  for column in range(columns)] for row in range(rows)]
    answer = []
    
    for y1, x1, y2, x2 in queries:
        x1 -= 1
        y1 -= 1
        x2 -= 1
        y2 -= 1
        
        change = []
        temp = rc[y1][x1]
        change.append(temp)
        for x in range(x1, x2):
            rc[y1][x + 1], temp = temp, rc[y1][x + 1]
            change.append(temp)
        
        for y in range(y1, y2):
            rc[y + 1][x2], temp = temp, rc[y + 1][x2]
            change.append(temp)
        
        for x in range(x2, x1, -1):
            rc[y2][x - 1], temp = temp, rc[y2][x - 1]
            change.append(temp)
        
        for y in range(y2, y1, -1):
            rc[y - 1][x1], temp = temp, rc[y - 1][x1]
            change.append(temp)

        answer.append(min(change))
        
                
    
    return answer